File: Go/Go_1000-/2383.go
// Go
// leetcode 2383
// https://leetcode.cn/problems/minimum-hours-of-training-to-win-a-competition/

func minNumberOfHours(initialEnergy int, initialExperience int, energy []int, experience []int) int {
	energySum, experienceTrained := 0, 0
	for _, e := range energy {
		energySum += e
	}

	curExp := initialExperience
	for _, exp := range experience {
		if curExp <= exp {
			experienceTrained += exp + 1 - curExp
			curExp = exp + 1
		}
		curExp += exp
	}
	
	if energySum >= initialEnergy {
		return experienceTrained + energySum + 1 - initialEnergy
	}
	return experienceTrained
}

File: cpp/cpp_1000-/2383.cc
// C++
// leetcode 2383
// https://leetcode.cn/problems/minimum-hours-of-training-to-win-a-competition/

class Solution {
public:
    int minNumberOfHours(int initialEnergy, int initialExperience, vector<int>& energy, vector<int>& experience) {
        int n = energy.size();
        int expTrain = 0;
        for (int i = 0; i < n; i++) {
        	initialEnergy -= energy[i];
        	if (initialExperience <= experience[i]) {
        		expTrain += experience[i] - initialExperience + 1;
        		initialExperience = experience[i] + 1;
        	}
        	initialExperience += experience[i];
        }
        return expTrain + (initialEnergy > 0 ? 0 : 1 - initialEnergy);
    }
};